src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,267 @@ PhaseStatus Compiler::optUnrollLoops()
+                    }
+                    asg_node = tree;
+                    asg_stmt = stmt;
+
+                    if (op2->gtOper == GT_SELECT)
+                    {
+                        found_select = true;
+                    }
+                    break;
+                }
+
+                // These do not need conditional execution.
+                case GT_NOP:
+                    break;
+
+                // Cannot optimise this block.
+                default:
+                    return false;
+            }
+        }
+    }
+    if (asg_node == nullptr)
+    {
+        // The blocks checked didn't contain any ASG nodes.
+        return false;
+    }
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("Attempting to conditionally execute " FMT_BB " from " FMT_BB "\n", middle_block->bbNum, block->bbNum);
+        for (BasicBlock* dump_block = block; dump_block != middle_block->bbNext; dump_block = dump_block->bbNext)
+        {
+            fgDumpBlock(dump_block);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    if (found_select)
+    {
+        // The assign is already conditional. Try adding another condition.
+        if (!optIfConvertCCmp(last->gtGetOp1(), asg_node))
+        {
+            return false;
+        }
+    }
+    else
+    {
+        // Make the assign conditional.
+        if (!optIfConvertSelect(last->gtGetOp1(), asg_node))
+        {
+            return false;
+        }
+    }
+
+    // Remove the JTRUE statement.
+    last->ReplaceWith(gtNewNothingNode(), this);
+    fgSetStmtSeq(block->lastStmt());
+    gtSetEvalOrder(last);
+    fgSetStmtSeq(asg_stmt);
+
+    // Update the flow.
+    fgRemoveAllRefPreds(block->bbJumpDest, block);
+    block->bbJumpKind = BBJ_NONE;
+    block->bbJumpDest = middle_block;
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("After if conversion\n");
+        for (BasicBlock* dump_block = block; dump_block != middle_block->bbNext; dump_block = dump_block->bbNext)
+        {
+            fgDumpBlock(dump_block);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    return true;
+#endif
+}
+
+bool Compiler::optIfConvertSelect(GenTree* original_condition, GenTree* asg_node)
+{
+    assert(original_condition->OperIsCmpCompare());
+    assert(asg_node->gtOper == GT_ASG);
+
+    // Duplicate the input of the assign.
+    // This will be used as the false result of the select node.
+    GenTree* current_value = gtCloneExpr(asg_node->AsOp()->gtOp1);
+    current_value->gtFlags &= GTF_EMPTY;
+
+    // Duplicate the condition and invert it
+    GenTree* cond = gtCloneExpr(original_condition);
+    cond->gtFlags |= GTF_DONT_CSE;
+    cond->gtFlags ^= GTF_RELOP_JMP_USED;
+    cond->gtFlags ^= GTF_RELOP_NAN_UN;
+    cond->gtOper = GenTree::ReverseRelop(cond->gtOper);
+
+    // Create a select node.
+    GenTreeConditional* select =
+        gtNewConditionalNode(GT_SELECT, cond, asg_node->AsOp()->gtOp2, current_value, asg_node->TypeGet());
```

Comment:
This is handling the case of:
```csharp
var x = ...;
if (b)
  x = ...;
```

I assume in the future it will be beneficial to handle also
```csharp
int x;
if (b)
  x = ...;
else
  x = ...;
```
?

Reply:
Yes, I'm not handling any `else` conditions in this PR.

Also, I'm not handling any `or` conditions either:
```
var x = ...;
if (b || c)
  x = ...;
```

That would be for a follow on patch